src/eg/src/verifiable_decryption.rs

```diff
@@ -0,0 +1,895 @@
+#![deny(clippy::unwrap_used)]
+#![deny(clippy::expect_used)]
+#![deny(clippy::panic)]
+#![deny(clippy::manual_assert)]
+
+//! This module provides the implementation of verifiable decryption for [`Ciphertext`]s.
+//! For more details see Section `3.6` of the Electionguard specification `2.0.0`.
+
+use crate::{
+    discrete_log::DiscreteLog,
+    election_parameters::ElectionParameters,
+    fixed_parameters::FixedParameters,
+    guardian::GuardianIndex,
+    guardian_public_key::GuardianPublicKey,
+    guardian_share::GuardianSecretKeyShare,
+    hash::{eg_h, HValue},
+    joint_election_public_key::{Ciphertext, JointElectionPublicKey},
+};
+use itertools::izip;
+use num_bigint::BigUint;
+use num_traits::{One, Zero};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use util::{
+    csprng::Csprng,
+    integer_util::{
+        get_single_coefficient_at_zero, group_lagrange_at_zero, mod_inverse, to_be_bytes_left_pad,
+    },
+};
+
+/// A decryption share is a guardian's partial decryption of a given ciphertext.
+///
+/// This corresponds to the `M_i` in Section `3.6.2`.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct DecryptionShare {
+    /// The guardian's index
+    pub i: GuardianIndex,
+    #[serde(
+        serialize_with = "util::biguint_serde::biguint_serialize",
+        deserialize_with = "util::biguint_serde::biguint_deserialize"
+    )]
+    /// The decryption share value
+    pub m_i: BigUint,
+}
+
+impl DecryptionShare {
+    /// This function computes the [`DecryptionShare`] for a given [`Ciphertext`] and [`GuardianSecretKeyShare`].
+    ///
+    /// The arguments are
+    /// - `self` - the encrypted share
+    /// - `fixed_parameters` - the fixed parameters
+    /// - `secret_key_share` - the secret key share
+    /// - `ciphertext` - the ElGamal ciphertext
+    pub fn from(
+        fixed_parameters: &FixedParameters,
+        secret_key_share: &GuardianSecretKeyShare,
+        ciphertext: &Ciphertext,
+    ) -> Self {
+        let m_i = ciphertext
+            .alpha
+            .modpow(&secret_key_share.p_i, fixed_parameters.p.as_ref());
+        DecryptionShare {
+            i: secret_key_share.i,
+            m_i,
+        }
+    }
+}
+
+/// The combined decryption share allows to compute the plain-text from a given ciphertext.
+///
+/// This corresponds to the `M` in Section `3.6.2`.
+#[derive(Debug)]
+pub struct CombinedDecryptionShare(BigUint);
+
+/// Represents errors occurring while combining [`DecryptionShare`]s into a [`CombinedDecryptionShare`].
+#[derive(Error, Debug, PartialEq)]
+pub enum ShareCombinationError {
+    /// Occurs if not enough shares were provided. Combination requires shares from at least `k` out of `n` guardians.
+    #[error("Only {l} decryption shares given, but at least {k} required.")]
+    NotEnoughShares { l: usize, k: u32 },
+    /// Occurs if multiple shares from the same guardian are provided.
```

Comment:
I think this documention line is a duplicate of the one below

Reply:
Fixed